### DISCUSSION TITLE
Import existing API definitions and convert to CRDs

### DISCUSSION THREAD

**Discussion by joshblakeley** [TykTechnologies/tyk-operator#263] [Ideas] [answered]
## Expected Behavior

Often users will have existing Tyk deployments either in different environments to those using the Operator or in existing Tyk deployments where the operator is to be onboarded. To move to the operator a user must manually create the CRDs which can be an onerous task.

We should be able to import an existing API definition and where possible template it onto a new CRD and pass that back. Preferably just as a CLI command for automation.

The dashboard may then import the package that does this work and export CRDs from our Cloud and On premise API Managers.

## Current Behavior

No real solution to this currently.

## Possible Solution

One option is to use Go templating to pass values from an inputted API definition JSON and create a CRD

**Comment by gernest** [comment]
**notes**

# CRD names

we can't directly use `name` field of api definition because this might conflict with resource names ( There is no restrictions on the api definitions , any string is valid)

# Linked API's

For looping, we should dismantle `tyk://` scheme url and construct corresponding looping targets object

# API ID
Whatever whatever is decided on the CRD name above, will be combined with the supplied namespace , so we completely ignore `api_id` from dash/gw

# Policies

The exports should be of both policies and api definitions to allow cross referencing of resources

.... etc

**Comment by asoorm** [comment, marked as answer]
Just a quick update - this is being addressed here: https://github.com/TykTechnologies/tyk-operator/pull/481
